core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -0,0 +1,247 @@
+        };
+    }
+
+
+    /**
+     * The target of the ObjectParser. One of these is built every time you call {@linkplain ObjectSpooler#apply(XContentParser, Object)}
+     * Note that it is not static so it inherits {@linkplain ObjectSpooler}'s type parameters.
+     */
+    public class Spool {
+        /**
+         * Array of constructor args to be passed to the {@link ObjectSpooler#builder}.
+         */
+        private final Object[] constructorArgs = new Object[constructorArgNames.size()];
+        private final XContentParser parser;
+        /**
+         * How many of the constructor parameters have we collected? We keep track of this so we don't have to count the
+         * {@link #constructorArgs} array looking for nulls when we receive another constructor parameter. When this is equal to the size of
+         * {@link #constructorArgs} we build the target object.
+         */
+        private int constructorArgsCollected = 0;
+        /**
+         * List of spooled attributes to be saved to the target object when we can build it. This is only allocated if an attribute has to
+         * be spooled.
+         */
+        private List<Consumer<Value>> spooled;
+        /**
+         * The target object. This will be instantiated with the constructor arguments are all parsed.
+         */
+        private Value targetObject;
+
+        public Spool(XContentParser parser) {
+            // TODO we probably don't need the parse here any more
+            this.parser = parser;
+        }
+
+        /**
+         * Set a constructor argument and build the target object if all constructor arguments have arrived.
+         */
+        private void constructorArg(int position, Object value) {
+            if (constructorArgs[position] != null) {
+                throw new IllegalArgumentException("Can't repeat constructor param");
+            }
+            constructorArgs[position] = value;
+            constructorArgsCollected++;
+            if (constructorArgsCollected == constructorArgNames.size()) {
+                try {
+                    unspool();
+                } catch (ParsingException e) {
+                    throw new ParsingException(e.getLineNumber(), e.getColumnNumber(),
+                            "[" + objectParser.getName() + "] failed to unspool fields", e);
+                } catch (Exception e) {
+                    throw new ParsingException(null, "[" + objectParser.getName() + "] failed to unspool fields", e);
+                }
+            }
+        }
+
+        /**
+         * Spool up a consumer that we'll call once the targetObject is built. If targetObject has been built just apply the consumer
+         * immediately.
+         */
+        private void spool(Consumer<Value> spoolMe) {
+            assert targetObject == null: "Don't spool after the targetObject has been built! Just apply the consumer directly.";
+            if (spooled == null) {
+                spooled = new ArrayList<>();
```

Comment:
Do we know how many things (worst case) we are going to spool at this point? Is it just the number of fields declared to the parser? If so, should we preinitialise the array list with that size so it doesn't need to resize during parsing?